cas-server-webapp-support/src/main/java/org/jasig/cas/util/CasLoggerContextInitializer.java

```diff
@@ -0,0 +1,164 @@
+
+    private ServletContextListener loggerContext;
+
+    private final String loggerContextPackageName;
+
+    private final Resource logConfigurationFile;
+
+    private final String logConfigurationField;
+
+    /**
+     * Instantiates a new Cas logger context initializer.
+     */
+    protected CasLoggerContextInitializer() {
+        this.loggerContext = null;
+        this.loggerContextPackageName = null;
+        this.logConfigurationField = null;
+        this.logConfigurationFile = null;
+    }
+
+    /**
+     * Instantiates a new Cas logger context initializer.
+     *
+     * @param loggerContextPackageName the logger context package name
+     * @param logConfigurationFile the log configuration file
+     * @param logConfigurationField the log configuration field
+     */
+    public CasLoggerContextInitializer(@NotNull final String loggerContextPackageName,
+                                       @NotNull final Resource logConfigurationFile,
+                                       @NotNull final String logConfigurationField) {
+        this.loggerContextPackageName = loggerContextPackageName;
+        this.logConfigurationField = logConfigurationField;
+        this.logConfigurationFile = logConfigurationFile;
+    }
+
+    /**
+     * Initialize the logger by decorating the context
+     * with settings for the log file and context.
+     * Calls the initializer of the logging framework
+     * to start the logger.
+     */
+    private void initialize() {
+        try {
+            if (!INITIALIZED.get() && this.loggerContext != null) {
+                final ServletContextEvent event = new ServletContextEvent(this.context);
+                this.loggerContext.contextInitialized(event);
+                LOGGER.debug("Initialized logging context via [{}]. Logs will be written to [{}]",
+                        this.loggerContext.getClass().getSimpleName(),
+                        this.logConfigurationFile);
+                INITIALIZED.set(true);
+            }
+        } catch (final Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Destroys all logging hooks and shuts down
+     * the logger.
+     */
+    @PreDestroy
+    public void destroy() {
+        try {
+            if (INITIALIZED.get() && this.loggerContext != null) {
+                final ServletContextEvent event = new ServletContextEvent(this.context);
+                LOGGER.debug("Destroying logging context and shutting it down");
+                this.loggerContext.contextDestroyed(event);
+            }
+        } catch (final Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Prepares the logger context. Locates the context and
+     * sets the configuration file.
+     * @return the logger context
+     */
+    private ServletContextListener prepareAndgetContextListener() {
+        try {
+            if (StringUtils.isNotBlank(this.loggerContextPackageName)) {
```

Comment:
Here is the longer explanation:

Log4j2 automatically starts itself via no configuration. It attempts to locate the log4j configuration file from a number of predefined places, if you dont set the location of the config file. If you want to explicitly control the location of the file, it is suggested that the setting be placed inside the app's web.xml file, which isn't going to work for us, because we want to let Spring handle the initialization and the setting of the config file for log4j and control that location from the cas.properties file.

Spring itself only works with log4j1. No support for v2 exists. Hence the reason for this class's existence.

This is what it does:

1. First, we disable the auto-initialization of log4j2 via the web.xml
2. Then, in the log4jConfiguration, we specify the path to the log4jConfiguration, and the package where the log4j2 initializer exists. This is the same initializer that would run automatically [which we disabled] and now we are going to find it and invoke it programmatically.
3. So, the code here attempts to locate the initializer, and set the log4jConfiguration file location parameter and then starts the logger context.